src/main/scala/com/twitter/finagle/Postgres.scala

```diff
@@ -0,0 +1,167 @@
+import com.twitter.util.{Duration, Return, Throw}
+import org.jboss.netty.channel.{ChannelPipelineFactory, Channels}
+
+object Postgres {
+
+  abstract class RequiredParam[T](name: String) extends Param[T] {
+    def default = throw new IllegalArgumentException(s"$name must be defined")
+  }
+
+  case class User(user: String) extends AnyVal
+  object User { implicit object param extends RequiredParam[User]("User") }
+
+  case class Password(password: Option[String]) extends AnyVal
+  object Password { implicit val param = Param(Password(None)) }
+
+  case class Database(database: String) extends AnyVal
+  object Database { implicit object param extends RequiredParam[Database]("Database") }
+
+  case class CustomTypes(types: Option[Map[Int, postgres.Client.TypeSpecifier]]) extends AnyVal
+  object CustomTypes { implicit val param = Param(CustomTypes(None)) }
+
+  case class CustomReceiveFunctions(functions: PartialFunction[String, ValueDecoder[T] forSome {type T}]) extends AnyVal
+  object CustomReceiveFunctions { implicit val param = Param(CustomReceiveFunctions(PartialFunction.empty)) }
+
+  case class BinaryResults(binaryResults: Boolean) extends AnyVal
+  object BinaryResults { implicit val param = Param(BinaryResults(false)) }
+
+  case class BinaryParams(binaryParams: Boolean) extends AnyVal
+  object BinaryParams { implicit val param = Param(BinaryParams(false)) }
+
+  private def defaultParams = StackClient.defaultParams +
+    ProtocolLibrary("postgresql") + Label("postgres") + FailFast(false) + param.ResponseClassifier(responseClassifier) +
+    Retries.Policy(retryPolicy)
+
+  private def defaultStack = StackClient.newStack[PgRequest, PgResponse]
+    .replace(StackClient.Role.prepConn, PrepConnection)
+
+  private def pipelineFactory(params: Stack.Params) = {
+    val Transport.TLSClientEngine(ssl) = params[Transport.TLSClientEngine]
+
+    new ChannelPipelineFactory {
+      def getPipeline = {
+        val pipeline = Channels.pipeline()
+
+        pipeline.addLast("binary_to_packet", new PacketDecoder(ssl.nonEmpty))
+        pipeline.addLast("packet_to_backend_messages", new BackendMessageDecoder(new BackendMessageParser))
+        pipeline.addLast("backend_messages_to_postgres_response", new PgClientChannelHandler(ssl, ssl.nonEmpty))
+        pipeline
+      }
+    }
+  }
+
+  private def mkTransport(params: Stack.Params) =
+    Netty3Transporter.apply[PgRequest, PgResponse](
+      pipelineFactory(params),
+      params + Transport.TLSClientEngine(None)  // we want to give this param to Postgres but not directly to transport
+                                                // because postgres doesn't start out in TLS
+    )
+
+  case class Client(
+    stack: Stack[ServiceFactory[PgRequest, PgResponse]] = defaultStack,
+    params: Stack.Params = defaultParams
+  ) extends StdStackClient[PgRequest, PgResponse, Client] {
+    type In = PgRequest
+    type Out = PgResponse
+
+    def newRichClient(): postgres.Client = {
```

Comment:
You can have `newRichClient` taking `Name` instead of grabbing it from stack params.